gwcs/wcs.py

```diff
@@ -634,3 +641,273 @@ def fix_inputs(self, fixed):
+                                                             u, v, undist_x, undist_y,
+                                                             ud, vd, undist_xd, undist_yd,
+                                                             verbose=verbose)
+        # The Following is necessary to put the fit into the SIP formalism.
+        cdmat, sip_poly_x, sip_poly_y = _reform_poly_coefficients(fit_poly_x, fit_poly_y)
+        # cdmat = np.array([[fit_poly_x.c1_0.value, fit_poly_x.c0_1.value],
+        #                   [fit_poly_y.c1_0.value, fit_poly_y.c0_1.value]])
+        det = cdmat[0][0] * cdmat[1][1] - cdmat[0][1] * cdmat[1][0]
+        U = ( cdmat[1][1] * undist_x - cdmat[0][1] * undist_y) / det
+        V = (-cdmat[1][0] * undist_x + cdmat[0][0] * undist_y) / det
+        detd = cdmat[0][0] * cdmat[1][1] - cdmat[0][1] * cdmat[1][0]
+        Ud = ( cdmat[1][1] * undist_xd - cdmat[0][1] * undist_yd) / detd
+        Vd = (-cdmat[1][0] * undist_xd + cdmat[0][0] * undist_yd) / detd
+  
+        if max_inv_pix_error:
+            fit_inv_poly_u, fit_inv_poly_v, max_inv_resid = _fit_2D_poly(ntransform,
+                                                            npoints, None,
+                                                            max_inv_pix_error,
+                                                            U, V, u-U, v-V,
+                                                            Ud, Vd, ud-Ud, vd-Vd,
+                                                            verbose=verbose)
+        pdegree = fit_poly_x.degree
+        if pdegree > 1:
+            hdr['a_order'] = pdegree
+            hdr['b_order'] = pdegree
+            _store_2D_coefficients(hdr, sip_poly_x, 'A')
+            _store_2D_coefficients(hdr, sip_poly_y, 'B')
+            if max_inv_pix_error:
+                _store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+                _store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+            hdr['sipmxerr'] = (max_resid * plate_scale, 'Max diff from GWCS (equiv pix).')
+            if max_inv_pix_error:
+                ipdegree = fit_inv_poly_u.degree
+                hdr['ap_order'] = ipdegree
+                hdr['bp_order'] = ipdegree
+                hdr['sipiverr'] = (max_inv_resid, 'Max diff for inverse (pixels)')
+        else:
+            hdr['ctype1'] = 'RA---TAN'
+            hdr['ctype2'] = 'DEC--TAN'
+
+        hdr['cd1_1'] = cdmat[0][0]
+        hdr['cd1_2'] = cdmat[0][1]
+        hdr['cd2_1'] = cdmat[1][0]
+        hdr['cd2_2'] = cdmat[1][1]
+        return hdr
+
+
+def _fit_2D_poly(ntransform, npoints, degree, max_error,
+                 xin, yin, xout, yout,
+                 xind, yind, xoutd, youtd,
+                 verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied transform.
+
+    Use adaptive sampling based on the current degree.
+    """
+    llsqfitter = LinearLSQFitter()
+
```

Comment:
Maybe it would be worth defining here "verbose print" - `vprint` and then use it everywhere in the code instead of the `if verbose: print("blah")`?
```suggestion

    def vprint(*args):
        if verbose:
            print(*args)
            
```

Reply:
This won't work, will it? We need to pass verbose too, right?

Reply:
Shouldn't it know what is the verbose value? It seems to work for me:
```python
In [45]: def f(verbose=True):
    ...:     def vprint(*args):
    ...:         if verbose:
    ...:             print(*args)
    ...:     vprint("Hello")
    ...:

In [46]: f(False)

In [47]: f()
Hello
```

Reply:
Looking at it it does seem that it is used only in one function. But, I'm not sure I like doing that. I'll think about it.

Reply:
I thought about it. Threshold isn't high enough to justify it imho.